shared/src/metabase/shared/util/parameters.cljc

```diff
@@ -0,0 +1,145 @@
+                  [metabase.shared.util.i18n :refer [trs]])]))
+
+(defn- formatted-list
+  [values]
+  (str (str/join ", " (butlast values)) " " (trs "and") " " (last values)))
+
+(defmulti formatted-value
+  "Formats a value appropriately for inclusion in a text card, based on its type. Does not do any escaping.
+  For datetime parameters, the logic here should mirror the logic (as best as possible) in
+  frontend/src/metabase/parameters/utils/date-formatting.ts"
+  (fn [tyype _value] (keyword tyype)))
+
+(defmethod formatted-value :date/single
+  [_ value]
+  #?(:cljs (.format (moment value) "MMMM D, YYYY")
+     :clj value))
+
+(defmethod formatted-value :date/month-year
+  [_ value]
+  #?(:cljs (let [m (moment value "YYYY-MM")]
+             (if (.isValid m) (.format m "MMMM, YYYY") ""))
+     :clj value))
+
+(defmethod formatted-value :date/quarter-year
+  [_ value]
+  #?(:cljs (let [m (moment value "[Q]Q-YYYY")]
+             (if (.isValid m) (.format m "[Q]Q, YYYY") ""))
+     :clj value))
+
+(defmethod formatted-value :date/range
+  [_ value]
+  (let [[start end] (str/split value "~")]
+    (if (and start end)
+      (str (formatted-value :date/single start)
+           " - "
+           (formatted-value :date/single end))
+      "")))
+
+(defmethod formatted-value :date/relative
+  [_ value]
+  (case value
+    "today"      (trs "Today")
+    "yesterday"  (trs "Yesterday")
+    "past7days"  (trs "Past 7 Days")
+    "past30days" (trs "Past 30 Days")
+    "lastweek"   (trs "Last Week")
+    "lastmonth"  (trs "Last Month")
+    "lastyear"   (trs "Last Year")
+    "thisday"    (trs "Today")
+    "thisweek"   (trs "This Week")
+    "thismonth"  (trs "This Month")
+    "thisyear"   (trs "This Year")
+    ;; Always fallback to default formatting, just in case
+    (formatted-value :default value)))
+
+(defmethod formatted-value :date/all-options
+  [_ value]
+  ;; Test value against a series of regexes (similar to those in metabase/parameters/utils/mbql.js) to determine
+  ;; the appropriate formatting, since it is not encoded in the parameter type.
+  ;; TODO: this is a partial implementation that only handles simple dates
+  (condp (fn [re value] (->> (re-find re value) second)) value
+    #"^~?([0-9-T:]+)~?$"       :>> (partial formatted-value :date/single)
+    #"^([0-9-T:]+~[0-9-T:]+)$" :>> (partial formatted-value :date/range)
+    (str value)))
+
+(defmethod formatted-value :default
+  [_ value]
+  (cond
+    (and (sequential? value) (> (count value) 1))
+    (formatted-list value)
+
+    (sequential? value)
+    (str (first value))
+
+    :else
+    (str value)))
+
+(def ^:private escaped-chars-regex
+  #"[\\/*_`'\[\](){}<>#+-.!$@%^&=|\?~]")
+
+(defn- escape-chars
+  [text]
+  (str/replace text escaped-chars-regex #(str \\ %)))
+
+(defn- replacement
+  [tag->param match]
+  (let [tag-name (second match)
+        param    (get tag->param tag-name)
+        value    (:value param)
+        tyype    (:type param)]
```

Comment:
i think you did this because it collides with the default `type`?
if so maybe change it to param-type or sth?

Reply:
Yeah, but the `tyype` convention is used elsewhere as well, eg: https://github.com/metabase/metabase/blob/nm-params-in-text/shared/src/metabase/types.cljc#L328-L329

I'm cool with changing to `param-type` though

Reply:
oh, if we already using it then I think it's okay.